### PR TITLE
Time function

### PR DESCRIPTION
**Pull request by realwebdev**
### check the model activity for time function review.

**I was trying to retrieve start and end save them to variable then apply Sub func from time library. wasn't successful. Please check this and let me know the mistake. I am unable to save st as start time variable as query output.
func TotalActivitytime(activity_id uint,  db *gorm.DB) {
    // activity := Activity{}
    // st := db.Where(map[string]interface{}{"ID": activity_id}).First(activity.Start_time)
    // if st != nil {
    //  log.Print("Error occured while updating activity end time!")
    //  return
    // }
    // et := db.Where(map[string]interface{}{"ID": activity_id}).First(activity.End_time)
    // if et != nil {
    //  log.Print("Error occured while updating activity end time!")
    //  return
    // }
    difference := et.Sub(st)
    tt := db.Table("activities").Where("ID = ?", activity_id).Updates(map[string]interface{}{"total_time": difference})
    if tt != nil {
        log.Print("Error occured while updating activity end time!")
        return
    }
    log.Print(difference)
}**

### CODE REVIEW

File: database/conn.go [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+	_ "github.com/jinzhu/gorm/dialects/postgres"
+)
+
+func AutoMigrate(modelObject interface{}, db *gorm.DB) error {
+	if err := db.DropTableIfExists(modelObject).Error; err != nil {
+		log.Print("Error occured While Deleting Previous Table")
+		return err
+	}
+	log.Print("Previous Table drop successfully")
+	if err := db.AutoMigrate(modelObject).Error; err != nil {
```
Comment (zunnorainzaheer) [inline database/conn.go:18]: There should be a line break above this, as this is a different code block.

Comment (Bilaltanveer38) [inline database/conn.go:18]: There should be a line break above this, as this is a different code block.

File: main.go [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+	// models.SignIn("kkskitchawn@gmail.com1", "mypass2", db)
+	// models.UserDeletion(1, db)
+
+	// models.CreateProject(models.Project{UserID: 3, Project_name: "MYFirstProj1"}, db)
+	// models.DeleteProject(2, db)
+	// models.UpdateProject(1, "newnameproject2", db)
+
+	// models.CreateStartActivity(models.Activity{Activity_name: "MYFirsActivity", ProjectID: 1, UserID: 1, Start_time: time.Now()}, db)
+	// models.EndActivity(1, time.Now(), db)
+	// models.TotalActivitytime(1, time.Now(), time.Now(), db)
```
Comment (Bilaltanveer38) [inline main.go:27]: not the right signature

File: models/activity.go [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+	log.Print("your activity has been created")
+}
+
+func EndActivity(activity_id uint, endtime interface{}, db *gorm.DB) {
+	if err := db.Table("activities").Where("ID = ?", activity_id).Updates(map[string]interface{}{"end_time": endtime}).Error; err != nil {
+		log.Print("Error occured while updating activity end time!")
+		return
+	}
+	log.Print("Successfully updated end of activity in DB")
+
```
Comment (Bilaltanveer38) [inline models/activity.go:34]: remove that space.

File: models/activity.go [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+func CreateStartActivity(activity Activity, db *gorm.DB) {
+	if err := db.Create(&activity).Error; err != nil {
+		log.Print("Error Occured while Create the activity in DB!")
+		return
+	}
+	log.Print("your activity has been created")
+}
+
+func EndActivity(activity_id uint, endtime interface{}, db *gorm.DB) {
+	if err := db.Table("activities").Where("ID = ?", activity_id).Updates(map[string]interface{}{"end_time": endtime}).Error; err != nil {
```
Comment (Bilaltanveer38) [inline models/activity.go:29]: you need to calculate the total time here as well, and update it with the end time.

File: models/activity.go [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+}
+
+func TotalActivitytime(activity_id uint, db *gorm.DB) {
+	// activity := Activity{}
+	// st := db.Where(map[string]interface{}{"ID": activity_id}).First(activity.Start_time)
+	// if st != nil {
+	// 	log.Print("Error occured while updating activity end time!")
+	// 	return
+	// }
+	// et := db.Where(map[string]interface{}{"ID": activity_id}).First(activity.End_time)
```
Comment (Bilaltanveer38) [inline models/activity.go:44]: not the correct syntax for retrieving the object from the database.

File: models/activity.go [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+		log.Print("Error occured while updating activity end time!")
+		return
+	}
+	log.Print("Successfully updated end of activity in DB")
+
+}
+
+func TotalActivitytime(activity_id uint, db *gorm.DB) {
+	// activity := Activity{}
+	// st := db.Where(map[string]interface{}{"ID": activity_id}).First(activity.Start_time)
```
Comment (Bilaltanveer38) [inline models/activity.go:39]: syntax is not correct.

It should be

if err:= db.Table("activities").Where("id=?", activity_id).First(&activity).Error; err!=nil{
   return err
}
startTime:=activity.start_time
endTime:=activity.end_time

File: models/activity.go [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+func EndActivity(activity_id uint, endtime interface{}, db *gorm.DB) {
+	if err := db.Table("activities").Where("ID = ?", activity_id).Updates(map[string]interface{}{"end_time": endtime}).Error; err != nil {
+		log.Print("Error occured while updating activity end time!")
+		return
+	}
+	log.Print("Successfully updated end of activity in DB")
+
+}
+
+func TotalActivitytime(activity_id uint, db *gorm.DB) {
```
Comment (Bilaltanveer38) [inline models/activity.go:37]: no need of the separate function, you can do this in EndTime func

File: models/activity.go [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+
+func DeleteActivity(activityid uint, db *gorm.DB) {
+	if err := db.Table("activities").Where("ID=?", activityid).Delete(Activity{}).Error; err != nil {
+		log.Print("Error occured activity does not exist in DB!")
+		return
+	}
+	log.Print("Activity successfully deleted from DB")
+}
+
+func UpdateActivity(activity_id uint, new_name interface{}, db *gorm.DB) {
```
Comment (Bilaltanveer38) [inline models/activity.go:66]: why passing new_name as interface{}?

Comment (realwebdev) [reply models/activity.go:66]: change to string.

File: models/activity.go (hunk 7 above)
Comment (realwebdev) [reply models/activity.go:37]: followed

**Pull request merged**